File: 프로그래머스/lv2/17683. ［3차］ 방금그곡/［3차］ 방금그곡.py
from collections import deque

def change(s): # 올림(#)이 있다면 소문자로 변환 후 반환
    temp = deque(s)
    s = []
    while temp:
        if temp[0] == '#':
            temp.popleft()
            s[-1] = s[-1].lower()
        else:s.append(temp.popleft())
    return "".join(s)

def solution(m, mif):
    answer = ''
    d = []
    m = change(m)
    for i in range(len(mif)):
        temp = mif[i].split(',')
        temp[-1] = change(temp[-1])
        t =  60*(int(temp[1][:2])-int(temp[0][:2])) + int(temp[1][3:5])-int(temp[0][3:5])
        rep = temp[-1]*2 # 2번 반복했을 때의 악보
        if t<2*len(rep):temp[-1] = rep[:t] # 완벽하게 2번 반복을 안했다면
        else:temp[-1] *= len(m)//len(temp[-1]) + 2
        if m in temp[-1]:d.append([t,len(mif)-i,temp[-2]])
    if not d:
        return '(None)'
    d = sorted(d,reverse=True)
    return d[0][-1]

File: 프로그래머스/lv2/17683. ［3차］ 방금그곡/［3차］ 방금그곡.java
import java.util.*;

class Solution {
    int toMinute(String time) {
        String[] temp = time.split(":");
        return Integer.parseInt(temp[0]) * 60 + Integer.parseInt(temp[1]);
    }
    String toLower(String s) {
        List<String> result = new ArrayList<>();
        String[] temp = s.split("");
        
        for (int i=0;i<s.length();i++) {
            if (temp[i].equals("#")) {
                result.set(result.size()-1, result.get(result.size()-1).toLowerCase());
            } else {
                result.add(temp[i]);
            }
        }
        return String.join("", result);
    }
    String getMelody(String lyrics, int t) {
        String result = "";
        int n = lyrics.length();
        int share = t / n, remainder = t % n;
        
        for (int i=0;i<share;i++) {
            result += lyrics;
        }
        result += lyrics.substring(0, remainder);
        return result;
    }
    public String solution(String m, String[] musicinfos) {
        int time = 0, t = 0;
        String answer = "(None)";
        String[] temp;
        String start, end, title, lyrics, melody;
        m = toLower(m);
        for (String musicinfo: musicinfos) {
            temp = musicinfo.split(",");
            start = temp[0];
            end = temp[1];
            title = temp[2];
            lyrics = toLower(temp[3]);
            
            t = toMinute(end) - toMinute(start);
            melody = getMelody(lyrics, t);
            
            if (melody.contains(m) && time < t) {
                time = t;
                answer = title;
            }
        }
        return answer;
    }
}